src/Uno.Toolkit.UI/Controls/SafeArea/SafeArea.cs

```diff
@@ -0,0 +1,541 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Text;
+using System.Threading.Tasks;
+using Windows.Foundation;
+using Windows.UI.ViewManagement;
+using Microsoft.Extensions.Logging;
+using Uno.Extensions;
+
+#if IS_WINUI
+using Microsoft.UI.Xaml;
+using Microsoft.UI.Xaml.Controls;
+using Microsoft.UI.Xaml.Input;
+using Microsoft.UI.Xaml.Media;
+using Microsoft.UI.Xaml.Media.Animation;
+using XamlWindow = Microsoft.UI.Xaml.Window;
+#else
+using Windows.UI.Xaml;
+using Windows.UI.Xaml.Controls;
+using Windows.UI.Xaml.Input;
+using Windows.UI.Xaml.Media;
+using Windows.UI.Xaml.Media.Animation;
+using XamlWindow = Windows.UI.Xaml.Window;
+#endif
+
+namespace Uno.Toolkit.UI
+{
+	public partial class SafeArea : ContentControl
+	{
+		[Flags]
+		public enum InsetMask
+		{
+			All = Left | Right | Top | Bottom,
+			None = 0,
+			Top = 1,
+			Bottom = 2,
+			Left = 4,
+			Right = 8,
+			SoftInput = 16
+		}
+
+		public enum InsetMode
+		{
+			Padding,
+			Margin
+		}
+
+#if HAS_UNO
+		private static readonly ILogger _log = typeof(SafeArea).Log();
+#endif
+
+		/// <summary>
+		/// The insets of the safe area relative to the entire XamlWindow.
+		/// </summary>
+		/// <remarks>This will be 0 if the entire window is 'safe' for content.</remarks>
+		public static Thickness WindowInsets => GetWindowInsets();
+
+		private static void OnInsetsChanged(DependencyObject dependencyObject, DependencyPropertyChangedEventArgs args)
+		{
+#if WINUI
+			// SafeArea is disabled for WinUI 3 and up as there's no available API for bounds.
+#else
+			if (dependencyObject is FrameworkElement fe)
+			{
+				SafeAreaDetails.GetInstance(fe).OnInsetsChanged((InsetMask)args.OldValue, (InsetMask)args.NewValue);
+			}
+			else
+			{
+#if HAS_UNO // Is building using Uno.UI
+				if (dependencyObject.Log().IsEnabled(LogLevel.Debug))
+				{
+					dependencyObject.Log().LogDebug($"Insets is only supported on FrameworkElement (Found {dependencyObject?.GetType()})");
+				}
+#endif
+			}
+#endif
+		}
+
+		private static void OnSafeAreaOverrideChanged(DependencyObject dependencyObject, DependencyPropertyChangedEventArgs args)
+		{
+#if WINUI
```

Comment:
Are we using `WINUI` to distinguish if we're building for Windows? We probably should be using `!HAS_UNO`, because WinUI is present in Uno as well.

Reply:
Yeah it should be checking if it is Windows WinUI. So !HAS_UNO.

I'm confused now, inside of Uno's VisiblesBoundPadding.cs we are using #if WinUI

It should still work for Uno platforms targeting WinUI, Just not Windows.

Reply:
@jeromelaban, I want to make sure that we basically do nothing if we are running on a WinAppSdk Windows app. This should work for all other platforms, including UWP. Is #if WINUI not the right condition here?